File: apps/player-linux/src/core/playback-defaults.test.ts
import { describe, expect, it } from "vitest";
import { resolvePlaybackItemSettings } from "./player";

describe("authoritative playback defaults", () => {
  it("overrides authored media values only when the item delegates", () => {
    const item = {
      id: "item",
      assetId: "asset",
      assetType: "image",
      durationMs: 10_000,
      fitMode: "contain",
      transition: "none",
      audioEnabled: true,
      volume: 1,
      deliveryPolicy: "download",
      usePlayerDefaults: true,
    };
    expect(
      resolvePlaybackItemSettings(
        item,
        {
          defaultFitMode: "cover",
          defaultTransition: "fade",
          defaultAudioEnabled: false,
          defaultVolume: 0.25,
        },
        22_000,
      ),
    ).toMatchObject({
      durationMs: 22_000,
      fitMode: "cover",
      transition: "fade",
      audioEnabled: false,
      volume: 0.25,
    });
  });

  it("preserves authored values when delegation is disabled", () => {
    const item = {
      id: "item",
      assetId: "asset",
      assetType: "image",
      durationMs: 10_000,
      fitMode: "stretch",
      transition: "crossfade",
      audioEnabled: true,
      volume: 0.9,
      deliveryPolicy: "download",
      usePlayerDefaults: false,
    };
    expect(
      resolvePlaybackItemSettings(item, { defaultFitMode: "cover" }, 22_000),
    ).toMatchObject({
      durationMs: 10_000,
      fitMode: "stretch",
      transition: "crossfade",
      audioEnabled: true,
      volume: 0.9,
    });
  });
});

File: apps/player-android/app/src/test/java/org/tilecast/player/content/PlaybackDefaultsTest.kt
package org.tilecast.player.content

import org.junit.Assert.assertEquals
import org.junit.Test
import org.tilecast.player.network.ManifestItem
import org.tilecast.player.network.PlayerPlaybackDefaults

class PlaybackDefaultsTest {
    private val defaults = PlayerPlaybackDefaults(
        defaultVolume = .25,
        defaultFitMode = "cover",
        defaultImageDurationSeconds = 22,
        defaultTransition = "fade",
        defaultAudioEnabled = false,
    )

    @Test fun delegatedItemUsesEveryPlaybackDefault() {
        val item = ManifestItem("item", "asset", "variant", "image", 10_000, "contain", "none", true, 1f, deliveryPolicy = "download", usePlayerDefaults = true)
        val resolved = item.withPlaybackDefaults(defaults)
        assertEquals(22_000L, resolved.durationMs)
        assertEquals("cover", resolved.fitMode)
        assertEquals("fade", resolved.transition)
        assertEquals(false, resolved.audioEnabled)
        assertEquals(.25f, resolved.volume)
    }

    @Test fun authoredItemRemainsAuthoritative() {
        val item = ManifestItem("item", "asset", "variant", "image", 10_000, "stretch", "crossfade", true, .9f, deliveryPolicy = "download", usePlayerDefaults = false)
        val resolved = item.withPlaybackDefaults(defaults)
        assertEquals(10_000L, resolved.durationMs)
        assertEquals("stretch", resolved.fitMode)
        assertEquals("crossfade", resolved.transition)
        assertEquals(true, resolved.audioEnabled)
        assertEquals(.9f, resolved.volume)
    }
}
